src/rdf4cpp/rdf/storage/node/NodeStorage.cpp

```diff
@@ -1,142 +1,281 @@
+     * and acquire-load the new value, we don't care about this value and will simply retry on the next slot.
+     *
+     * If we win the race we store the pointer with a release-store in which
+     * case the compare_exchange of all other threads racing for this slot will fail.
+     *
+     * Storing the refcount with a relaxed-store is fine here since there can be no
+     * other threads racing for the same reference count at this point. This is
+     * because there can currently be no other NodeStorages pointing to this slot, since by definition this
+     * execution is the one that creates the first one pointing to this slot in this generation.
+     * Additionally existing WeakNodeStorages pointing to this slot are not interested
+     * in it as the generation increases here with a release-store which is synchronized with Weak::upgrade.
+     *
+     * see https://marabos.nl/atomics/memory-ordering.html#example-lazy-initialization-with-indirection
+     */
+    for (size_t ix = 0; ix < NodeStorage::node_context_instances.size(); ++ix) {
+        auto &slot = NodeStorage::node_context_instances[ix];
 
-    if (stored_instance != backend_instance and stored_instance != nullptr)
-        throw std::runtime_error("A NodeStorage with manager_id " + std::to_string(backend_instance->manager_id.value) + " is already registered.");
+        INodeStorageBackend *old_value = nullptr;
+        if (slot.backend.compare_exchange_strong(old_value, backend_instance, std::memory_order_release, std::memory_order_acquire)) {
+            // found slot
+            slot.generation.fetch_add(1, std::memory_order_release); // release for synchronization with Weak::upgrade
+            slot.refcount.store(1, std::memory_order_relaxed);
 
-    node_context_instances[backend_instance->manager_id.value] = backend_instance;
-    return NodeStorage(backend_instance);
+            return NodeStorage{identifier::NodeStorageID{static_cast<uint16_t>(ix)}};
+        } else if (old_value == backend_instance) {
+            throw std::runtime_error{"The provided backend is already registered"};
+        }
+    }
+
+    throw std::runtime_error{"Maximum number of backend instances exceeded"};
 }
+
 void NodeStorage::unregister_backend(INodeStorageBackend *backend_instance) {
-    if (backend_instance == nullptr)
+    if (backend_instance == nullptr) {
         throw std::runtime_error("Backend instance must not be null.");
-    node_context_instances[backend_instance->manager_id.value] = nullptr;
+    }
+
+    auto it = std::find_if(NodeStorage::node_context_instances.begin(), NodeStorage::node_context_instances.end(), [&](auto const &slot) {
+        return slot.backend == backend_instance;
+    });
+
+    if (it == NodeStorage::node_context_instances.end()) {
+        throw std::runtime_error{"Backend instance not found"};
+    }
+
+    // TODO: just remove this function? It is basically just always UB when the reference count
```

Comment:
Opinion?

Reply:
The idea behind unregister_backend was to have a way to prevent a call tothe constructor of a persistent INodeStorage implementation. There might also be other ways.

Reply:
Not 100% sure what you mean by that. Do you mean persistent as in Tentris' node storage or something different?
And how would it prevent that ctor call?

Reply:
Sorry, I meant destructor. Yes, like in Tentris.

Reply:
Reviewing the current design of tentris' persistent node storage, I would say we already have sth. like that with the polymorphic implementation of INodeStorage and it's non-polymorphic implementation that is available via a pointer. 

So no, I don't think we need it anymore